Me: Improve UI when photo and Gravatar profile are hidden

danielbachhuber opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/63199

## Proposed Changes

### Clarifies language on the "hide Gravatar" toggle

Instead of saying "Hide Gravatar profile", say "Hide photo and Gravatar profile":

#### After

<img width="754" alt="image" src="https://user-images.githubusercontent.com/36432/176886851-22268aeb-2616-4f26-ad4c-9ba402a3a55c.png">

#### Before

<img width="739" alt="image" src="https://user-images.githubusercontent.com/36432/176682911-68fc7c61-e694-4987-bdcb-3954f91a09c3.png">

### Visually indicates when Gravatar is hidden

When the user's photo and Gravatar profile are hidden, the page appears as follows:

<img width="761" alt="image" src="https://user-images.githubusercontent.com/36432/176886954-9dfdaca5-cd89-4299-a90c-2c4d4b64643a.png">

<img width="853" alt="image" src="https://user-images.githubusercontent.com/36432/176886989-be09774a-c250-4e29-94a3-c919aa6facab.png">

When the user's photo and Gravatar are enabled, the page appears as follows:

<img width="775" alt="image" src="https://user-images.githubusercontent.com/36432/176887025-f03c7135-98ca-45a4-ae2a-012759efd9f2.png">

#### Testing Instructions

1. Navigate to `/me`.
2. Observe the page appears as documented above.
3. Toggle the "Hide photo and Gravatar profile" button.
4. Observe the changed display of the page.

danielbachhuber committed:
Me: Clarify that the toggle hides both the photo and profile

danielbachhuber committed:
Me: Visually indicate when photo and Gravatar profile is hidden

danielbachhuber committed:
Merge branch 'trunk' into update/me-edit-gravatar-ui-hidden

danielbachhuber commented:
@Automattic/build One thing that's kinda funky: it takes a brief moment for `userSettings` to load. This causes the Gravatar photo to display briefly when refreshing the page, until the `gravatar_profile_hidden` setting is set to `true`:

https://user-images.githubusercontent.com/36432/176684253-01c4c786-2c4f-452a-894d-44b5fab4065c.mp4

Any ideas on how to work around this?

danielbachhuber committed:
Allow `border-radius: 50%;` to prevent the linter from erroring

danielbachhuber committed:
Clean up JS linting issues

wojtekn commented:
@danielbachhuber feature-wise looks great to me.

> This causes the Gravatar photo to display briefly when refreshing the page

I also observed this behavior. I tried to quickly refresh and click on the upload button before it disappears, but with no luck. This is good, but that blinking is weird indeed.

Maybe we should have another state, like a placeholder image (similar to one used when the photo is disabled), and return this one if `userSettings` are not loaded yet?

p-jackson commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -254,6 +300,7 @@ const recordReceiveImageEvent = () => recordTracksEvent( 'calypso_edit_gravatar_
 export default connect(
 	( state ) => ( {
 		user: getCurrentUser( state ) || {},
+		userSettings: getUserSettings( state ),
```
Using the `getUserSetting` selector instead means that the component will update immediately when the user toggles rather than having to wait until the user clicks "Save changes". (that's because this selector consults the unsaved settings too).

It also has the advantage that it will only rerender the component when the setting we're interested in changes, not when _any_ setting changes.

```suggestion
		isGravatarProfileHidden: getUserSetting( state, 'gravatar_profile_hidden' ),
```

p-jackson replied:
Just realised you'd still need to use the `getUserSettings()` selector in order to decide whether the settings were currently loading and the placeholders should be rendered. But personally, I'd still use the singular `getUserSetting()` to get the boolean from the store, since it feels much nicer to have the Gravatar component update in real-time as you toggle the setting.

p-jackson commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -153,11 +154,56 @@ export class EditGravatar extends Component {
 		// it's the popular value for large Gravatars in Calypso
 		const GRAVATAR_IMG_SIZE = 400;
+
+		if ( userSettings.gravatar_profile_hidden ) {
```
It's too bad that this hidden profile version has to involve so much copy-paste, but I think I can see how it's _just_ different enough from the public profile version that it'd be just as messy to try and merge the two cases together but using if/elses to change what's needed. 😮‍💨

danielbachhuber replied:
In the future, we'll be able to write code with our minds, and we won't even have to worry about which file the code goes in!

p-jackson reviewed (commented):
> it takes a brief moment for `userSettings` to load

Yes for me this is taking a couple of seconds to load so it does look pretty bad. I like the idea of using placeholders too. The "Profile links" section on this page is also using them so it should feel fine. Not intending to increase the scope of this PR even further, but the toggle switch suffers from the same issue so maybe it should get a rectangle placeholder too, or at least render as disabled while user settings are loading?

You can use this mixin https://github.com/Automattic/wp-calypso/blob/a2d8ff3b0590cb08034bc310fd00abd8b594fff7/client/assets/stylesheets/shared/mixins/_placeholder.scss#L6 to add the standard Calypso pulsing animation to your placeholder elements.

p-jackson commented on client/me/profile/index.jsx:
```diff
@@ -120,7 +120,7 @@ class Profile extends Component {
 								onChange={ this.toggleGravatarHidden }
 								label={ this.props.translate(
-									'{{spanLead}}Hide my Gravatar profile.{{/spanLead}} {{spanExtra}}This will prevent your {{profileLink}}Gravatar profile{{/profileLink}} and photo from appearing on any site. It may take some time for the changes to take effect. Gravatar profiles can be deleted at {{deleteLink}}Gravatar.com{{/deleteLink}}.{{/spanExtra}}',
+									'{{spanLead}}Hide my photo and Gravatar profile.{{/spanLead}} {{spanExtra}}This will prevent your {{profileLink}}Gravatar profile{{/profileLink}} and photo from appearing on any site. It may take some time for the changes to take effect. Gravatar profiles can be deleted at {{deleteLink}}Gravatar.com{{/deleteLink}}.{{/spanExtra}}',
```
I sometimes like to be a little extra when deploying a string change like this where it's a block of text that's only slightly changed.

As soon as this deploys, because we don't have the translations for it yet, it's going to default to showing the English version for all non-English speaking users. Since the new message is only incrementally better than the old message, Spanish speakers for instance would be better off keeping the old message until translations are ready. It'd also be a big chunk of unreadable text appearing in a quite important part of the UI, so it might be a bit unnerving not knowing what it says!

```js
import i18n, { localize } from 'i18n-calypso';
import { englishLocales } from '@automattic/i18n-utils';

function MyComponent( { translate, locale } ) {
	const message = englishLocals.includes( locale ) && i18n.hasTranslation( 'The updated string' )
		? translate( 'The updated string' )
		: translate( 'The old string' );

	return <p>{ message }</p>;
}

// I'll use the localize HoC here to get access to `translate` and `locale` props
// since that's how this profile view works.
export default localize( MyComponent );
```

Whenever a translation becomes available in glotpress, it'll appear in the UI. Once all the mag16 languages are translated we can tidy up this code as a Janitorial task.

---

The simpler way of doing this is to simply not merge the PR until all translations are complete. This is doable with the `String Freeze` label. p4TIVU-5Jq-p2 (don't apply the label until you're fairly certain you have the final wording)
But that might be a few weeks, and we presumably don't want to block this bug from being fixed for that long.

danielbachhuber replied:
> I sometimes like to be a little extra when deploying a string change like this where it's a block of text that's only slightly changed.

@p-jackson Great suggestion! I've applied an implementation of your example in 2600366850

danielbachhuber committed:
Merge branch 'trunk' into update/me-edit-gravatar-ui-hidden

danielbachhuber committed:
Use the placeholder mixin to add a loading state to `EditGravatar`

danielbachhuber commented:
@p-jackson Added the loading UX to the avatar with c4dde83a861c522a83a934b27bbdbe9124c3859d

https://user-images.githubusercontent.com/36432/176773828-b34b2de5-e35f-4a25-9a72-d4d5944333c9.mp4

Is this the best way to determine whether the selector is loading, or is there a more precise approach?

https://github.com/Automattic/wp-calypso/pull/65122/commits/c4dde83a861c522a83a934b27bbdbe9124c3859d#diff-3d48acc81a692f5fa88e3c076767498266c4f5fbea40435394e30be798152fb2R171

danielbachhuber committed:
Use the old Gravatar string until the new one has a locale translation

danielbachhuber committed:
Add loading UX to the Gravatar toggle element

danielbachhuber commented:
Added loading UX to the toggle fieldset too in 4343b5eda52a91b3063e82054aff99d881d6277d

https://user-images.githubusercontent.com/36432/176781691-c56fc8d2-c585-4177-a31e-9c025f07d9da.mp4

clucasrowlands commented:
Hello from the Publications Team (formerly Editorial)! 

Copy-wise, looks fine. You could even swap the order so that it's consistent with the wording just after, but that's a small thing, and ultimately up to you.

> **Hide my Gravatar profile and photo.** This will prevent your Gravatar profile and photo from appearing...

danielbachhuber commented:
@clucasrowlands Thanks!

How would you feel about swapping the other instance?

> **Hide my photo and Gravatar profile.** This will prevent your photo and Gravatar profile from appearing...

The reason I suggest this is because it seems like 'photo' is the primary object we're discussing.

clucasrowlands commented:
> How would you feel about swapping the other instance?

Works for me!

wojtekn commented:
> Is this the best way to determine whether the selector is loading, or is there a more precise approach?

I see in a few places we use `isUpdatingUserSettings` from `calypso/state/user-settings/selectors`, maybe this one would work?

wojtekn commented:
@danielbachhuber I see one thing that may need an adjustment - the width of the photo container changes when we enable/disable photo, and it causes name fields to be slightly moving:

![Screen Capture on 2022-07-01 at 11-30-41](https://user-images.githubusercontent.com/727413/176868841-5996fbe1-bdb7-40cc-9893-cc50043260e2.gif)

danielbachhuber committed:
Tweak language the prioritize 'photo'

danielbachhuber committed:
Use the more precise `isFetchingUserSettings`

danielbachhuber committed:
Use a `min-width` to avoid slight jiggle when toggling settings

danielbachhuber commented:
> > How would you feel about swapping the other instance?
> 
> Works for me!

@clucasrowlands Great! Updated in 0c34a9ec1e

> I see in a few places we use `isUpdatingUserSettings` from `calypso/state/user-settings/selectors`, maybe this one would work?

@wojtekn `isUpdatingUserSettings` doesn't work (only `true` when _updating_)... but `isFetchingUserSettings` works great!

Thanks for the tip 😄 Cleaned up with b4930f3098

> I see one thing that may need an adjustment - the width of the photo container changes when we enable/disable photo, and it causes name fields to be slightly moving:

Good catch! Cleaned that up in 27951a48e9

https://user-images.githubusercontent.com/36432/176892386-808c550d-7e4b-48f6-9af5-2430edc760e4.mp4

p-jackson commented on client/blocks/edit-gravatar/test/index.jsx:
```diff
@@ -23,6 +23,10 @@ describe( 'EditGravatar', () => {
 		email_verified: false,
 	};
 
+	const userSettings = {
```
This `userSettings` prop can be removed from all the tests now. Most of the test cases are just dealing with situations where `isGravatarProfileHidden` and `isFetchingUserSettings` are falsey, so they're working just fine without the new props having to be explict.

p-jackson reviewed (approved):
> `isFetchingUserSettings` works great!

Perfect!

I've retested the component in `/me` and in the P2 signup process at `/start/p2` and it's working great.

Left a note about cleaning up the unit tests, but this is looking good to go. I think it's a great improvement 👍

zaguiini commented on client/me/profile/updated-gravatar-string.js:
```diff
@@ -0,0 +1,32 @@
+		};
+		return englishLocales.includes( locale ) ||
+			i18n.hasTranslation(
+				'{{spanLead}}Hide my photo and Gravatar profile.{{/spanLead}} {{spanExtra}}This will prevent your photo and {{profileLink}}Gravatar profile{{/profileLink}} from appearing on any site. It may take some time for the changes to take effect. Gravatar profiles can be deleted at {{deleteLink}}Gravatar.com{{/deleteLink}}.{{/spanExtra}}'
+			)
+			? translate(
+					'{{spanLead}}Hide my photo and Gravatar profile.{{/spanLead}} {{spanExtra}}This will prevent your photo and {{profileLink}}Gravatar profile{{/profileLink}} from appearing on any site. It may take some time for the changes to take effect. Gravatar profiles can be deleted at {{deleteLink}}Gravatar.com{{/deleteLink}}.{{/spanExtra}}',
+					stringParts
+			  )
```
Could we extract these strings, that are identical, to a constant?

danielbachhuber replied:
> Could we extract these strings, that are identical, to a constant?

@zaguiini Unfortunately, I tried that, but `translate()` doesn't like to receive a constant.

zaguiini commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -153,11 +155,70 @@ export class EditGravatar extends Component {
 		const GRAVATAR_IMG_SIZE = 400;
+
+		if ( this.props.isFetchingUserSettings ) {
+			return (
```
I like to keep the `render` method as slim as possible for edge cases. Would it be cleaner if we extracted those edge-case markups to a method and then call it here?

p-jackson replied:
That's true. Having the `render` method mostly just contain the happy-path UI is nice.

danielbachhuber replied:
> I like to keep the `render` method as slim as possible for edge cases. Would it be cleaner if we extracted those edge-case markups to a method and then call it here?

@zaguiini Good suggestion! I did so with c3281ed5c9

zaguiini commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -153,11 +155,70 @@ export class EditGravatar extends Component {
+		}
+
+		if ( isGravatarProfileHidden ) {
+			return (
```
I like to keep the `render` method as slim as possible for edge cases. Would it be cleaner if we extracted those edge-case markups to a method and then call it here?

danielbachhuber replied:
Did with c3281ed5c9

zaguiini commented on client/me/profile/updated-gravatar-string.js:
```diff
@@ -0,0 +1,32 @@
+					stringParts
+			  )
+			: translate(
+					'{{spanLead}}Hide my Gravatar profile.{{/spanLead}} {{spanExtra}}This will prevent your {{profileLink}}Gravatar profile{{/profileLink}} and photo from appearing on any site. It may take some time for the changes to take effect. Gravatar profiles can be deleted at {{deleteLink}}Gravatar.com{{/deleteLink}}.{{/spanExtra}}',
```
This one too. It takes a lot of space and it's not clear why it exists.

danielbachhuber replied:
See above ^ — `translate()` doesn't want to receive a constant.

zaguiini reviewed (approved):
Functionality-wise, it works fine. I left some non-blocking comments on code organization.

danielbachhuber committed:
Merge branch 'trunk' into update/me-edit-gravatar-ui-hidden

danielbachhuber committed:
`userSettings` constant no longer necessary after b4930f3098482171ec941c11347d593a8c90ebde

danielbachhuber committed:
Abstract early render paths to dedicated functions

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7419597

Thank you @danielbachhuber for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7419597).

p-jackson commented:
PR to tidy up the `hasTranslation` check: #65626
